kafka-connect-s3/src/main/java/io/confluent/connect/s3/format/avro/AvroRecordWriterProvider.java

```diff
@@ -105,4 +166,102 @@ public void close() {
       }
     };
   }
+
+  /**
+   * Get output schema in avro format
+   *
+   * @param keySchema the connect schema of the key
+   * @param valueSchema the connect schema of the value
+   * @return output schema in avro format
+   */
+  private org.apache.avro.Schema mergeSchema(Schema keySchema, Schema valueSchema) {
+    // Create the connect schema builder
+    SchemaBuilder builder = SchemaBuilder
+            .struct()
+            .name(valueSchema.name())
+            .doc(valueSchema.doc())
+            .version(valueSchema.version());
+
+    // Process the key schema
+    if (keySchema.type() == Schema.Type.STRUCT) {
+      for (Field field: keySchema.fields()) {
+        builder.field(field.name(), field.schema());
+      }
+    } else if (supportedSchemaTypes.contains(keySchema.type())) {
+      builder.field("key", keySchema);
+    } else {
+      throw new ConnectException(
+              "Key type: " + keySchema.type().toString() + "not supported!"
+      );
+    }
+
+    // Process the value schema
+    if (valueSchema.type() == Schema.Type.STRUCT) {
+      for (Field field: valueSchema.fields()) {
+        builder.field(field.name(), field.schema());
+      }
+    } else if (supportedSchemaTypes.contains(valueSchema.type())) {
+      builder.field("value", valueSchema);
+    } else {
+      throw new ConnectException(
+              "Value type: " + valueSchema.type().toString() + "not supported!"
+      );
+    }
+
```

Comment:
Have we done any performance measurements? I'd hate to have this slow us down significantly.

Reply:
Agreed - need to get some real numbers on this all. We can mitigate this with parallelism... I think. Right @johnnycaol ? We can run multiple connector sinks per partition?

Reply:
@bellemare according to kafka documentation 
```
Kafka does it better. By having a notion of parallelism—the partition—within the topics, Kafka is able to provide both ordering guarantees and load balancing over a pool of consumer processes. This is achieved by assigning the partitions in the topic to the consumers in the consumer group so that each partition is consumed by exactly one consumer in the group. By doing this we ensure that the consumer is the only reader of that partition and consumes the data in order. Since there are many partitions this still balances the load over many consumer instances. Note however that there cannot be more consumer instances in a consumer group than partitions.
```

Kafka connect is the same, you can't have # of consumers more than partitions, however, you can have multiple connect instances for the same topic (maximum=# of topic partitions)

@jfb-flipp will test this.